### PR TITLE
Add support for configuration cache

### LINKED ISSUE #160: Support config-cache
When I run my application with `config-cache`. I receive warns that `gradle-git-properties` doesn't support it. 
image
It seems like [Undeclared reading of environment properties](https://docs.gradle.org/6.8.3/userguide/configuration_cache.html#config_cache:requirements:undeclared_env_var_read).

Do you think you can resolve it to support configuration cache?

Status of community plugins:
https://github.com/gradle/gradle/issues/13490

### PR DESCRIPTION
**Pull request by joschi**
Remove problematic invocations in the task action and replace them with the recommended replacements from the Gradle documentation.

https://docs.gradle.org/6.8.3/userguide/configuration_cache.html
https://docs.gradle.org/6.8.3/userguide/lazy_configuration.html

Closes #160

### CODE REVIEW

**Review by joschi — COMMENTED** [review]
@n0mer @tha2015 This PR contains the changes from #162 and #163 too, so depending on whether you want to merge those PRs, I can adapt this PR accordingly.

**Comment by joschi** [comment]
@tha2015 I've rebased the PR on the latest master branch and included the last missing part from #158.

Let me know if you have any questions about the changes in this PR. 😄

**Comment by tha2015** [comment]
Hi @joschi , sorry for the delay. I was busy and could not spend much time to review and test the changes for backward compatibility. Could you help to verify that it works on Gradle 4.x and also increase the plugin version? Then I will release it. Thanks.
 - Thai

**Comment by joschi** [comment]
> Could you help to verify that it works on Gradle 4.x

The changes won't work on Gradle 4.x because of some changes regarding how to build instances of `DirectoryProperty`.
- https://docs.gradle.org/4.10/javadoc/org/gradle/api/file/DirectoryProperty.html
- https://docs.gradle.org/5.6/javadoc/org/gradle/api/file/DirectoryProperty.html

This being said, I've made the plugin work with Gradle 5.1 up to Gradle 7.0-rc1 (see `BackwardCompatibilityFunctionalTest`) with a few exemptions from Gradle 6.0-6.3 because of a bug in Gradle's handling of properties (see comments in test).

Gradle 4.10.3 has been released on Dec 05, 2018 (more than two years ago).
Should this plugin still support Gradle 4.x?

**Comment by tha2015** [comment]
Thanks for the info. Agreed that we don't have to support it anymore. We will just need to change the plugin's major version and put a note so people will be able about this change.

**Comment by joschi** [comment]
@tha2015 I've update the backward-compatibility test to skip tests on unsupported Java versions, bumped the plugin version to 2.3.0, and added a compatibility matrix in the README.

File: README.md [hunk 1]
```diff
@@ -7,22 +7,31 @@ Idea - @lievendoclo, originally published in article [Spring Boot's info endpoin
...
 
 ```groovy
 plugins {
-  id "com.gorylenko.gradle-git-properties" version "2.2.4"
+  id "com.gorylenko.gradle-git-properties" version "2.3.0"
 }
 ```
 
-A git.properties file will be generated when building Java-based projects (the plugin will configure any existing `classes` task to depend on `generateGitProperties` task - which is responsible for generated git.properties file). For non-Java projects, `generateGitProperties` task must be executed explicitly to generate git.properties file. The git repository for the project will be used.
+A `git.properties file will be generated when building Java-based projects (the plugin will configure any existing `classes` task to depend on `generateGitProperties` task - which is responsible for generated `git.properties` file). For non-Java projects, `generateGitProperties` task must be executed explicitly to generate `git.properties` file. The git repository for the project will be used.
```
Comment (tha2015) [inline README.md:34]: Thanks for updating README document, it is really helpful. Still there are some places you forgot to end the code tags (see the README at https://github.com/joschi/gradle-git-properties/tree/config-cache )

**Comment by joschi** [comment]
@tha2015 Thanks for checking! I think I've caught them all now.

File: README.md [hunk 2]
```diff
@@ -92,7 +101,7 @@ git.remote.origin.url
 git.tags
 git.total.commit.count
 ```
-You can have more fine grained control of the content of 'git.properties' using `keys`:
+You can have more fine-grained control of the content of `git.properties using `keys`:
```
Comment (makigumo) [inline README.md:104]:
```suggestion
You can have more fine-grained control of the content of `git.properties` using `keys`:
```

Comment (joschi) [reply README.md]: Thanks!

**Comment by tha2015** [comment]
FYI, I've published it but it is not approved yet.

>Task :publishPlugins
>Publishing plugin com.gorylenko.gradle-git-properties version 2.3.0
>Thank you. Your new plugin com.gorylenko.gradle-git-properties has been submitted for approval by Gradle engineers. The >request should be processed within the next few days, at which point you will be contacted via email.

**Comment by joschi** [comment]
@tha2015 Thanks for merging! 😄 

Could we tag the release in this repository?
The latest tag is still https://github.com/n0mer/gradle-git-properties/releases/tag/v.2.2.4.

**Comment by tha2015** [comment]
Thanks, I've created the tag.

**Pull request merged**